File: src/main/python/SingleLinkedList.py
from typing import Optional


class _Node:
    def __init__(self, value):
        self.value = value
        self.next: Optional[_Node] = None


class SingleLinkedList:
    __head = _Node(value=None)
    __size = 0

    def get_size(self):
        return self.__size

    def append(self, value):
        if self.__size == 0:
            node = self.__head
        else:
            node = self.__get_node_at(index=self.__size - 1)

        node.next = _Node(value=value)

        self.__size += 1

    def append_left(self, value):
        node = _Node(value=value)

        node.next = self.__head.next
        self.__head.next = node

        self.__size += 1

    def insert(self, index, value):
        if index == 0:
            self.append_left(value=value)
        elif index == self.__size:
            self.append(value=value)
        else:
            prev_node = self.__get_node_at(index=index - 1)
            node = _Node(value=value)

            node.next = prev_node.next
            prev_node.next = node

            self.__size += 1

    def get(self, index):
        return self.__get_node_at(index=index).value

    def remove_at(self, index):
        self.__check_index(index=index)

        if index == 0:
            prev_node = self.__head
        else:
            prev_node = self.__get_node_at(index=index - 1)

        node = prev_node.next
        prev_node.next = node.next

        self.__size -= 1

        return node.value

    def __get_node_at(self, index):
        self.__check_index(index=index)

        node = self.__head
        for _ in range(index + 1):
            node = node.next

        return node

    def __check_index(self, index):
        if not (0 <= index < self.__size):
            raise IndexError

File: src/main/typescript/SingleLinkedList.ts
class _Node<T> {
    _value: T
    _next: _Node<T>

    constructor(value: T) {
        this._value = value
    }
}

export class SingleLinkedList<T> {
    __head: _Node<T> = new _Node<T>(null)
    __size: number = 0

    getSize() {
        return this.__size
    }

    append(value: T) {
        let node: _Node<T>
        if (this.__size === 0) {
            node = this.__head
        } else {
            node = this.__getNodeAt(this.__size - 1)
        }

        node._next = new _Node(value)

        this.__size++
    }

    appendLeft(value: T) {
        const node: _Node<T> = new _Node<T>(value)

        node._next = this.__head._next
        this.__head._next = node

        this.__size++
    }

    insert(index: number, value: T) {
        if (index === 0) {
            this.appendLeft(value)
        } else if (index === this.__size) {
            this.append(value)
        } else {
            const prevNode: _Node<T> = this.__getNodeAt(index - 1)
            const node: _Node<T> = new _Node(value)

            node._next = prevNode._next
            prevNode._next = node

            this.__size++
        }
    }

    get(index: number) {
        return this.__getNodeAt(index)._value
    }

    removeAt(index: number) {
        this.__checkIndex(index)

        let prevNode: _Node<T>
        if (index === 0) {
            prevNode = this.__head
        } else {
            this.__getNodeAt(index - 1)
        }

        const node = prevNode._next
        prevNode._next = node._next

        this.__size--

        return node._value
    }

    __getNodeAt(index: number) {
        this.__checkIndex(index)

        let node: _Node<T> = this.__head
        for (let i = 0; i < index + 1; i++) {
            node = node._next
        }

        return node
    }

    __checkIndex(index: number) {
        if (index < 0 || this.__size <= index) {
            throw new RangeError()
        }
    }
}
